vdslib/src/vespa/vdslib/state/clusterstate.cpp

```diff
@@ -280,34 +280,40 @@ ClusterState::getNodeCount(const NodeType& type) const
     return _nodeCount[type];
 }
 
+namespace {
+    NodeState _G_defaultSDState(NodeType::STORAGE, State::DOWN);
+    NodeState _G_defaultDDState(NodeType::DISTRIBUTOR, State::DOWN);
+    NodeState _G_defaultSUState(NodeType::STORAGE, State::UP);
+    NodeState _G_defaultDUState(NodeType::DISTRIBUTOR, State::UP);
+    [[noreturn]] void throwUnknownType(const Node & node) __attribute__((noinline));
+    void throwUnknownType(const Node & node) {
+        throw vespalib::IllegalStateException("Unknown node type " + node.getType().toString(), VESPA_STRLOC);
+    }
+}
+
 const NodeState&
 ClusterState::getNodeState(const Node& node) const
 {
-        // If beyond node count, the node is down.
-    if (node.getIndex() >= _nodeCount[node.getType()]) {
-        if (node.getType() == NodeType::STORAGE) {
-            static NodeState defaultSDState(NodeType::STORAGE, State::DOWN);
-            return defaultSDState;
-        } else if (node.getType() == NodeType::DISTRIBUTOR) {
-            static NodeState defaultDDState(NodeType::DISTRIBUTOR, State::DOWN);
-            return defaultDDState;
-        }
-        throw vespalib::IllegalStateException(
-                "Unknown node type " + node.getType().toString(), VESPA_STRLOC);
-    }
-        // If it actually has an entry in map, return that
+    // If it actually has an entry in map, return that
     std::map<Node, NodeState>::const_iterator it = _nodeStates.find(node);
     if (it != _nodeStates.end()) return it->second;
+
+    // If beyond node count, the node is down.
+    if (__builtin_expect(node.getIndex() >= _nodeCount[node.getType()], false)) {
```

Comment:
Is it cheaper in practice to always do the map lookup than to have this check prior to it?

Reply:
It layed out the code in a more efficient way. There seemed to be less jumps back and forth.
The most important change was moving the throwing out to separate non inlined method as it eliminated the need for pushing and popping on the stack.
The second one was reorganising to only call throwUnknownType in one place.
And then I assumed it is more likely the node is in the _nodeStates map, than being beyond the node count.